installinator/src/peers.rs

```diff
@@ -264,37 +325,70 @@ impl PeersImpl for ExamplePeers {
+        artifact_id: ArtifactId,
         sender: mpsc::Sender<Result<Bytes, progenitor_client::Error>>,
-        // ExamplePeers doesn't need cancel_receiver (yet)
-        _cancel_receiver: oneshot::Receiver<()>,
+        cancel_receiver: oneshot::Receiver<()>,
     ) {
-        let artifact_id = artifact_id.clone();
-        tokio::spawn(async move {
-            let res = if peer == Ipv6Addr::LOCALHOST {
-                let mut bytes = BytesMut::new();
-                bytes.extend_from_slice(b"example");
-                bytes.extend_from_slice(artifact_id.to_string().as_bytes());
-                Ok(bytes.freeze())
-            } else if peer == Ipv6Addr::UNSPECIFIED {
-                Err(progenitor_client::Error::ErrorResponse(
-                    ResponseValue::new(
-                        (),
-                        StatusCode::NOT_FOUND,
-                        Default::default(),
-                    ),
-                ))
-            } else {
-                panic!("invalid peer, unknown to ExamplePeers: {peer}")
-            };
+        // TODO: be able to fetch from sled-agent clients as well
+        let artifact_client = ArtifactClient::new(peer, &self.log);
 
-            // Ignore errors in case the channel is dropped.
-            _ = sender.send(res).await;
+        tokio::spawn(async move {
+            tokio::select! {
+                _ = artifact_client.fetch_artifact(artifact_id, sender) => {},
+                _ = cancel_receiver => {},
+            }
         });
     }
 }
 
+#[derive(Debug)]
+struct ArtifactClient {
+    log: slog::Logger,
+    client: wicketd_client::Client,
+}
+
+impl ArtifactClient {
+    fn new(addr: SocketAddrV6, log: &slog::Logger) -> Self {
+        let endpoint = format!("http://[{}]:{}", addr.ip(), addr.port());
+        let log = log.new(
+            slog::o!("component" => "ArtifactClient", "peer" => addr.to_string()),
+        );
+        let client = wicketd_client::Client::new(&endpoint, log.clone());
+        Self { log, client }
+    }
+
+    async fn fetch_artifact(
+        &self,
+        artifact_id: ArtifactId,
+        sender: mpsc::Sender<Result<Bytes, progenitor_client::Error>>,
+    ) {
+        let artifact_bytes = match self
+            .client
+            .get_artifact(&artifact_id.name, &artifact_id.version)
+            .await
+        {
+            Ok(artifact_bytes) => artifact_bytes,
+            Err(error) => {
+                // TODO: does this lose too much info wicketd_client::types::Error?
+                let _ = sender.send(Err(error.into_untyped())).await;
+                return;
+            }
+        };
+
+        slog::debug!(
+            &self.log,
+            "preparing to receive {:?} bytes from artifact",
+            artifact_bytes.content_length(),
+        );
+
+        let mut bytes = artifact_bytes.into_inner_stream();
+        while let Some(item) = bytes.next().await {
+            _ = sender.send(item.map_err(Into::into)).await;
```

Comment:
Nitpick, but why do we expect the receiver to remain open for the duration of this function?

- If we don't expect that: Shouldn't we bail on error?
- If we do expect that: Should we assert it?

Reply:
Good question! There's two separate tasks here so the receiver staying open isn't guaranteed. Guess we should just exit in that case.

Reply:
(We can also get rid of the cancel oneshot channel in that case, I think! Dropping the receiver indicates that the task should be cancelled.)

Reply:
Done.